File: solutions/python/circular-buffer/1/circular_buffer.py
class BufferFullException(BufferError):

    def __init__(self, message):
        self.message = message

class BufferEmptyException(BufferError):
    
    def __init__(self, message):
        self.message = message

class CircularBuffer:
    circular_buffer=[]
    head=0 #the newest data
    tail=0 #the oldest data
    capacity=0 # the number of used slots
    bufSize=0 # size of buffer
    isBufferFull=False
    
    def __init__(self, size):
        self.circular_buffer = [None] * size
        self.bufSize=size
        self.capacity=0
        self.head = 0
        self.tail = 0
        
    def read(self):
        if self.capacity==0:
            raise BufferEmptyException("Circular buffer is empty")
        else:
            value=self.circular_buffer[self.tail]
            self.circular_buffer[self.tail]=None
            self.capacity=self.capacity-1
            print(self.head,self.tail,self.capacity,self.circular_buffer)
            self.tail=self.tail+1
            if self.tail==self.bufSize:
                self.tail=0
            return value
 
    def write(self, data):
        if self.capacity==self.bufSize:
            raise BufferFullException("Circular buffer is full")
        else:
            self.circular_buffer[self.head]=data
            self.head=self.head+1
            if self.head==self.bufSize:
                self.head=0
            self.capacity=self.capacity+1
            
    def overwrite(self, data):
        if self.capacity==self.bufSize:
            self.circular_buffer[self.tail]=data
            self.tail=self.tail+1
            if self.tail==self.bufSize:
                self.tail=0
            self.head=self.head+1
            if self.head==self.bufSize:
                self.head=0
        else:
            self.write(data)

    def clear(self):
        if self.capacity>0:
            self.circular_buffer[self.tail]=None
            self.tail=self.tail+1
            if self.tail==self.bufSize:
                self.tail=0
            self.capacity=self.capacity-1
        

File: solutions/scala/circular-buffer/1/src/main/scala/CircularBuffer.scala
class EmptyBufferException() extends Exception {}

class FullBufferException() extends Exception {}

class CircularBuffer(var capacity: Int) {
  var buffer = new Array[Int](capacity)
  var head : Int = 0
  var tail : Int = 0
  var usedSlots : Int = 0
  def write(value: Int) = {
    if(usedSlots==capacity){
      throw new FullBufferException()
    }
    else{
      buffer(head)=value
      head=head+1
      if(head==capacity){
        head=0
      }
      usedSlots=usedSlots+1  
    }
  }

  def read(): Int = {
    if(usedSlots==0){
      throw new EmptyBufferException()
    }
    else{
      var value : Int = 0
      value=buffer(tail)
      buffer(tail)=0
      tail=tail+1
      if(tail==capacity){
        tail=0
      }
      usedSlots=usedSlots-1
      return value
    }
  }

  def overwrite(value: Int) = {
    if(usedSlots==capacity){
      buffer(tail)=value
      tail=tail+1
      if(tail==capacity){
        tail=0
      }
      head=head+1
      if(head==capacity){
        head=0
      }
    }
    else{
      write(value)
    }
  }

  def clear() = {
    if(usedSlots > 0){
      buffer(tail)=0
      tail=tail+1
      if(tail==capacity){
        tail=0
      }
      usedSlots=usedSlots-1
    }
  }
}